### PR TITLE
scheduler: seed random shuffle of nodes with eval ID

### PR DESCRIPTION
**Pull request by tgross**
Processing an evaluation is nearly a pure function over the state
snapshot, but we randomly shuffle the nodes. This means that
developers can't take a given state snapshot and pass an evaluation
through it and be guaranteed the same plan results.

But the evaluation ID is already random, so if we use this as the seed
for shuffling the nodes we can greatly reduce the sources of
non-determinism. Unfortunately golang map iteration uses a global
source of randomness and not a goroutine-local one, but arguably
if the scheduler behavior is impacted by this, that's a bug in the
iteration.

---

Reviewers: I've grabbed a random set of folks to look at this but maybe consider this mostly a "working proposal". I'm not certain there aren't side-effects I haven't considered here. But this could have been useful while debugging a large customer's incident recently.

### CODE REVIEW

File: scheduler/util.go [hunk 1]
```diff
@@ -376,11 +377,18 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
 	return out, nil
 }
 
-// shuffleNodes randomizes the slice order with the Fisher-Yates algorithm
-func shuffleNodes(nodes []*structs.Node) {
+// shuffleNodes randomizes the slice order with the Fisher-Yates
+// algorithm. We seed the random source with the eval ID (which is
+// random) to aid in postmortem debuggging of specific evaluations and
+// state snapshots.
```
Comment (tgross) [inline scheduler/util.go:383]: @schmichael noted (elsewhere):
> Will this cause the node iteration order to be the same every time we retry processing the eval? Or would we maintain the RNG state between retries so that each retry would shuffle based on the "next" random value?

It'd be the same seed on each retry. There's not an easy way to thread the retry state down, but given that the plan has to make progress to be retried (I think?), we could use the counts on the current iteration of the plan to "wiggle" the seed we get from the eval ID deterministically too.

File: scheduler/util.go [hunk 2]
```diff
@@ -376,11 +377,18 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
...
+func shuffleNodes(evalID string, nodes []*structs.Node) {
+
+	seed, _ := binary.Varint([]byte(evalID))
```
Comment (shoenig) [inline scheduler/util.go:386]: I don't think this is technically correct

> Varint decodes an int64 from buf and returns that value and the number of bytes read (> 0).

I think we only get 1 byte of randomness: 
https://go.dev/play/p/OttVV5vI_Xu

Though TBH that's probably fine ~

Comment (tgross) [reply scheduler/util.go:386]: Huh! I knew that the 2nd return value was the bytes read, but I didn't realize it'd only read one byte. But now I'm trying to figure out what circumstance it could ever return a negative number for overflow. The docs for that method are weird.


In any case, I also just realized that if we were to do https://github.com/hashicorp/nomad/issues/11779 this would result in all the evals effectively having the same seed for a long stretch of time. I'll update this to pull the bytes off the _end_ of the eval ID (which are random even if we do #11779) and use the `binary.BigEndian.Uint32` functions which appear to read as many bytes as you ask for.

File: scheduler/util.go [hunk 3]
```diff
@@ -376,11 +377,29 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
...
+	// if we have sortable IDs
+	buf := []byte(plan.EvalID)
+	seed := int64(binary.BigEndian.Uint32(buf[len(buf)-4:]))
+
+	// include the plan progress so that we don't retry with
+	// the exact same shuffle
+	seed = seed |
+		int64(len(plan.NodeAllocation)) |
+		int64(len(plan.NodeUpdate)) |
+		int64(len(plan.DeploymentUpdates))
```
Comment (schmichael) [inline scheduler/util.go:396]: If we place N allocs per attempt won't we end up with the same order? I'm still not sure that's a problem though. :thinking: 

Node reshuffling between attempts may workaround issues like #9506, and on attempts >1 these slices might all have 0 length resulting in not escaping a scheduler/planner disagreement. That's a lot of bad luck at once, and on the bright side this work + #11793 give us an incredible opportunity to fix otherwise intractable bugs.

So I'm **+1 shipping this as-is** if no one else can think of a more pathological condition _not_ reshuffling on retries could cause. If we are worried here are a couple options:

## Option 1: RefreshIndex

On partial application, the planner sets [`PlanResult.RefreshIndex`](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/plan_apply.go#L386-L392) which, in the absence of #9506 style bugs, should always increase between retries. Even with bugs like #9506 I would expect the Raft index to increase, but I don't think there's any guarantee.

`seed |= RefreshIndex` == `seed` when `RefreshIndex` is 0, so I think you could change this code to:

```suggestion
	// include the plan result's RefreshIndex so that we don't retry with
	// the exact same shuffle
	seed |= planResult.RefreshIndex
```

...once you plumb planResult all the way through... which might get ugly.

## Option 2: Retry Number
The retry number would be a nice as AFAICT it does not repeat for a specific eval, so `evalID ||  retry` should be unique per [`process()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/generic_sched.go#L219) > [`computePlacements()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/generic_sched.go#L473-L488) > [`shuffleNodes()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/stack.go#L72-L73)

[`retryMax()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/util.go#L315-L319) could be trivially altered to send the `attempts` int to the callback func (is `func()` now but would become `func(attempts int)`).

Adding a metric for attempts when `process()` succeeds would be trivial as well. I don't think we have a metric for `how many attempts is it taking evals to get processed?` and if it's >1 attempt that's an indication of scheduler contention. If it's occurring persistently operators may want to lower `num_schedulers` as they're just throwing CPU cycles away with plans that end up conflicting.

**Review by schmichael — APPROVED** [review]

**Comment by schmichael** [comment]
Oops, forgot to mention a test that asserts `shuffleNodes` is stable would be nice.

**Review by motobrowning — COMMENTED** [review]
Ok

File: scheduler/util.go [hunk 4]
```diff
@@ -376,11 +377,29 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
 	return out, nil
 }
 
-// shuffleNodes randomizes the slice order with the Fisher-Yates algorithm
-func shuffleNodes(nodes []*structs.Node) {
+// shuffleNodes randomizes the slice order with the Fisher-Yates
+// algorithm. We seed the random source with the eval ID (which is
+// random) to aid in postmortem debuggging of specific evaluations and
```
[the remark this answers is not in the document]
Comment (tgross) [reply scheduler/util.go:382]: 🤦

File: scheduler/util.go (hunk 3 above)
Comment (tgross) [reply scheduler/util.go:396]: I really like option 1 here. It shouldn't be too hard to plumb that through and it's a better indicator than we'd get from counts, especially because of:

> and on attempts >1 these slices might all have 0 length resulting in not escaping a scheduler/planner disagreement. 

> Adding a metric for attempts when `process()` succeeds would be trivial as well. I don't think we have a metric for `how many attempts is it taking evals to get processed?` and if it's >1 attempt that's an indication of scheduler contention.

Hm... I think that can be right at the top of `process`? Would we do that as a Sample type so that operators can get a histogram over it?

**Comment by tgross** [comment]
> Oops, forgot to mention a test that asserts `shuffleNodes` is stable would be nice.

Good call.

File: scheduler/util.go [hunk 5]
```diff
@@ -376,11 +377,26 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
...
+func shuffleNodes(plan *structs.Plan, nodes []*structs.Node) {
+
+	// use the last 4 bytes because those are the random bits
+	// if we have sortable IDs
+	buf := []byte(plan.EvalID)
+	seed := int64(binary.BigEndian.Uint32(buf[len(buf)-4:]))
+
+	// include the plan result's RefreshIndex so that we don't retry with
+	// the exact same shuffle
+	seed |= planResult.RefreshIndex
```
Comment (shoenig) [inline scheduler/util.go:393]: I think you want to use xor (`^`); plain or could result in a noop, e.g.
https://go.dev/play/p/NCrGoS4d0TK

Comment (schmichael) [reply scheduler/util.go]: Seth wins the Bit Twiddling Award.

File: scheduler/stack.go [hunk 6]
```diff
@@ -70,7 +70,8 @@ type GenericStack struct {
 
 func (s *GenericStack) SetNodes(baseNodes []*structs.Node) {
 	// Shuffle base nodes
-	shuffleNodes(baseNodes)
+	idx, _ := s.ctx.State().LatestIndex()
```
Comment (schmichael) [inline scheduler/stack.go:73]: I can't believe how many words I'm writing about `we need a nonce` :sweat_smile:, however I think there are some subtle pros and cons to this approach:

**tl;dr** This allows writing reproducible tests against deterministic state stores which is the main improvement we're looking for! :shipit: 

Pros:

1. Subsequent scheduling attempts will be shuffled differently! This accomplishes that goal perfectly. :tada:
2. Test cases with deterministic state store updates will make shuffle nodes deterministically.

Cons:

1. When trying to reproduce scheduling behavior from a user's snapshot, the `LatestIndex()` will be the snapshot's index. The snapshot was likely taken long (in Raft index terms (pun not intended)) after when the scheduling attempt we wish to reproduce was made. Obviously in cases like this the entire state snapshot is likely divergent from the case we wish to inspect, so we can only _reduce_ non-determinism, not remove it. I'm unsure how much value there is in _reducing_ non-determinism in this case.
2. Workers only ever enforce the snapshots are **>=** a Raft index (Eval.WaitIndex initially, the RefreshIndex on subsequent attempts). Whether the index is `==` (reproducible shuffle!) or `>` (oh no) is entirely dependent on races between the Raft/fsm and scheduling subsystems. Even if the indexes are only off-by-1 the nodes will be shuffled completely differently.

Comment (tgross) [reply scheduler/stack.go:73]: Given that the primary use case for this is to make it so that we can have deterministic tests, maybe the right way to deal with this is to just have the index used for the shuffle get logged somewhere. That way we can feed that index to the shuffle function as part of tests?

Comment (schmichael) [reply scheduler/stack.go:73]: Yeah that or the retry # option.

Although I don't think there's a rush as this should work for those tests where we either bypass raft and manually insert objects into the state store *or* we load a snapshot into an otherwise inactive server and perform some scheduling attempts.
